Add realm setting for enabling read receipts.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First two commits are for backend work of the setting.
- Next commit adds migration to set the setting for existing realms.
- Next two commits are for adding the setting and UI and tooltip for user setting.
- Next commit fixes the label for user setting.
- Last commit adds UI for realm-default of read-receipts setting.

Fixes: <!-- Issue link, or clear description.--> #22657

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![Screenshot from 2022-08-10 20-24-46](https://user-images.githubusercontent.com/35494118/183937311-c1f54a05-ee8f-477f-94d6-7aae0fc9f652.png)
![Screenshot from 2022-08-10 20-24-56](https://user-images.githubusercontent.com/35494118/183937361-3ec05e60-17dc-4e26-aa76-7ced5ed3e6bb.png)
![Screenshot from 2022-08-10 20-24-38](https://user-images.githubusercontent.com/35494118/183937393-36dadcf8-7477-466b-a4db-a4e10b2baa9a.png)

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

timabbott commented on zerver/migrations/0405_set_default_for_enable_read_receipts.py:
```diff
@@ -0,0 +1,24 @@
+) -> None:
+    Realm = apps.get_model("zerver", "Realm")
+    # We enable read receipts by default in realms which require an invitation to join.
+    Realm.objects.filter(invite_required=True).update(enable_read_receipts=True)
```
An organization is not open if `emails_restricted_to_domains || invite_required`; so I think we want;
`Realm.objects.exclude(invite_required=False).exclude(emails_restricted_to_domains=False).update(enable_read_receipts=True)`

for this query.

timabbott commented on zerver/models.py:
```diff
@@ -684,6 +684,8 @@ class Realm(models.Model):
 
     default_code_block_language: Optional[str] = models.TextField(null=True, default=None)
 
+    enable_read_receipts: bool = models.BooleanField(default=False)
```
Let's add a comment for this one -- "Whether read receipts are enabled in the organization. If disabled, they will not be available regardless of users' personal settings."

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -11498,6 +11504,14 @@ paths:
+                        description: |
+                          Present if `realm` is present in `fetch_event_types`.
+
+                          Whether read receipts is enabled in the organization or not.
```
We should add another sentence: "If disabled, read receipt data will be unavailable to clients, regardless of individual users' personal read receipt settings."

timabbott commented:
I posted a batch of comments; otherwise this lgtm.

sahil839 commented on static/templates/settings/organization_settings_admin.hbs:
```diff
@@ -142,6 +142,18 @@
+                            Enable read receitps (<i>Users can always disable their personal read receipts.</i>)
+                        {{/tr}}
+                    </label>
+                </div>
```
We need the text in parens to be in italics. And we pass a string for label to `settings_checkbox`. I am not sure if we can do it in any other way.

timabbott replied:
OK, that makes sense. Probably we could extend `settings_checkbox` template to support an `html_label` alternative field as a refactor to simplify this; seems better to avoid having duplicates of it scattered around.

sahil839 commented:
Updated the PR with all the changes asked above.
- Added comment for the field in `models.py`.
- Updated description in API docs.
- Udpated the migration and it now uses `Q` or construction method.

timabbott commented:
Fixed a label typo (`receitps`) and then merged this, thanks @sahil839!
